### PR TITLE
perf: Address performance of EthGetTransactionCount 

### PR DESCRIPTION
**Pull request by fridrik01**
Fixes: https://github.com/filecoin-project/lotus/issues/10538

**Context**
We have observed that `EthGetTransactionCount` is one of the hotspots on Glif production notes, and we are seeing regular 10-20 second latencies when calling this rpc method. We have also been able to replicate this issue on local nodes on our devboxes on mainnet when calling this method repeatedly.

**The fix**
I tracked the high latency spikes and they were correlated when we were running ExecuteTipSet while following the chain.

To address this, we should not rely on tipset computation to get nounce and instead look at the parent tipset and then count the messages sent from the 'addr' as suggested by steb.

**Test plan**
I started and synced my lotus node on mainnet and started calling  `EthGetTransactionCount` rpc method in a loop.

The latency spikes seemed to be all gone, which I then confirmed by looking at the `http://localhost:1234/debug/metrics` for the `getStateNonce` metric (getnonce_ms)

image

Of almost 35k calls, all were within 8ms (before this fix, there were often outliers in the 5+ second range)

### CODE REVIEW

File: chain/messagepool/provider.go [hunk 1]
```diff
@@ -77,6 +78,10 @@ func (mpp *mpoolProvider) PubSubPublish(k string, v []byte) error {
 	return mpp.ps.Publish(k, v) // nolint
 }
 
+func (mpp *mpoolProvider) GetActorBefore(addr address.Address, ts *types.TipSet) (*types.Actor, error) {
+	return mpp.sm.LoadActor(context.TODO(), addr, ts)
```
Comment (fridrik01) [inline chain/messagepool/provider.go:82]: TODO: Support IsLite

**Comment by Stebalien** [comment]
I can confirm that this fixes the performance issue (or, at least, we don't get 10s pauses).

File: chain/messagepool/messagepool.go [hunk 2]
```diff
@@ -1078,14 +1078,42 @@ func (mp *MessagePool) getStateNonce(ctx context.Context, addr address.Address,
...
+	if err != nil {
+		return 0, err
+	}
+
+	// get the nonce from the actor before ts
+	actor, err := mp.api.GetActorBefore(addr, ts)
+	if err != nil {
+		return 0, err
+	}
+	highestNounce := actor.Nonce
```
Comment (raulk) [inline chain/messagepool/messagepool.go:1091]:
```suggestion
	highestNonce := actor.Nonce
```

File: chain/messagepool/messagepool.go [hunk 3]
```diff
@@ -1078,14 +1078,42 @@ func (mp *MessagePool) getStateNonce(ctx context.Context, addr address.Address,
...
+
+		if maddr == raddr {
+			if msg.Nonce+1 > highestNounce {
+				highestNounce = msg.Nonce + 1
+			}
+		}
+	}
 
-	mp.nonceCache.Add(nk, act.Nonce)
+	mp.nonceCache.Add(nk, highestNounce)
```
Comment (raulk) [inline chain/messagepool/messagepool.go:1114]: Would prefer to call it nextNonce.

File: chain/messagepool/messagepool.go [hunk 4]
```diff
@@ -1078,14 +1078,42 @@ func (mp *MessagePool) getStateNonce(ctx context.Context, addr address.Address,
...
+
+		maddr, err := mp.resolveToKey(ctx, msg.From)
+		if err != nil {
+			log.Warnf("failed to resolve message from address: %s", err)
+			continue
+		}
+
+		if maddr == raddr {
+			if msg.Nonce+1 > highestNounce {
+				highestNounce = msg.Nonce + 1
```
Comment (raulk) [inline chain/messagepool/messagepool.go:1109]:
```suggestion
			if n := msg.Nonce+1; n > highestNounce {
				highestNounce = n
```

File: chain/messagepool/messagepool.go (hunk 3 above)
Comment (raulk) [inline chain/messagepool/messagepool.go:1114]: Would suggest to clean up here and rename this cache to stateNonceCache.

File: chain/messagepool/messagepool.go [hunk 5]
```diff
@@ -371,7 +371,7 @@ func (ms *msgSet) toSlice() []*types.SignedMessage {
 func New(ctx context.Context, api Provider, ds dtypes.MetadataDS, us stmgr.UpgradeSchedule, netName dtypes.NetworkName, j journal.Journal) (*MessagePool, error) {
 	cache, _ := lru.New2Q[cid.Cid, crypto.Signature](build.BlsSignatureCacheSize)
 	verifcache, _ := lru.New2Q[string, struct{}](build.VerifSigCacheSize)
-	noncecache, _ := lru.New[nonceCacheKey, uint64](256)
+	noncecache, _ := lru.New[nonceCacheKey, uint64](32768) // 32k * ~200 bytes = 6MB
```
Comment (raulk) [inline chain/messagepool/messagepool.go:374]: Does it make sense to impose this memory tax on every node, when only hosted RPC nodes would truly benefit as they are public deployments? Besides that, I'm not convinced this increase is useful since the cache will be practically stale every 30 seconds? (The most popular usage pattern is getting the nonce of the pending tipset)

File: chain/messagepool/messagepool.go [hunk 6]
```diff
@@ -1078,14 +1078,42 @@ func (mp *MessagePool) getStateNonce(ctx context.Context, addr address.Address,
 		return n, nil
 	}
 
-	act, err := mp.api.GetActorAfter(addr, ts)
+	raddr, err := mp.resolveToKey(ctx, addr)
```
Comment (maciejwitowski) [inline chain/messagepool/messagepool.go:1081]: `raddr` is only used much later in the `for` loop. It was suspicious for me that we assigned `raddr` but then kept on using `addr` in `GetActorBefore`. Wdyt about moving this line just before `for`?

File: chain/messagepool/provider.go [hunk 7]
```diff
@@ -58,6 +60,23 @@ func (mpp *mpoolProvider) IsLite() bool {
 	return mpp.lite != nil
 }
 
+func (mpp *mpoolProvider) getActorLite(addr address.Address, ts *types.TipSet) (*types.Actor, error) {
```
Comment (maciejwitowski) [inline chain/messagepool/provider.go:63]: It's a bit easier to follow for me if refactoring like this (this only extracts the function without any functional changes, right?) is done in a separate commit.

**Review by maciejwitowski — APPROVED** [review]
One more hotspot down! 💪

**Review by Stebalien — APPROVED** [review]
I agree with https://github.com/filecoin-project/lotus/pull/10700/files#r1173643973, but it's not a high priority.

We can definitely optimize this more as discussed (e.g., iterating over messages once per tipset and caching the results) but this fixes the main performance issue.

**Comment by fridrik01** [comment]
@raulk Updated test plan with results from concurrent stress test, overall looks good!
